lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -602,6 +602,143 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
+//  %5 = witness_method $@opened P, #P.foo!1 : <Self
+//     where Self : P> (Self) -> () -> Int, %4 :  $@opened P :
+//     $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@guaranteed
+//     τ_0_0) -> Int
+//  %6 = apply %5<@opened P>(%4) : $@convention(witness_method:
+//     P) <τ_0_0 where τ_0_0 : P> (@guaranteed τ_0_0) -> Int
+//  TO
+//  %4 = open_existential_ref %3 : $P to $@opened P
+//  %7 = unchecked_ref_cast %4 : $@opened P to $C
+//  %8 = class_method %7 : $C, #C.foo!1 : (C) -> () -> Int,
+//     $@convention(method) (@guaranteed C) -> Int
+//  %9 = apply %8(%7) : $@convention(method)
+//     (@guaranteed C) -> Int return %6 : $Int
+SILInstruction *
+SILCombiner::propagateSoleConformingType(FullApplySite Apply,
+                                         WitnessMethodInst *WMI) {
+  // If WMI has concrete conformance, it can be optimized.
+  if (WMI->getConformance().isConcrete())
+    return nullptr;
+
+  // If the lookup type is an opened existential type,
+  // it cannot be made concrete.
+  if (!WMI->getLookupType()->isOpenedExistential())
+    return nullptr;
+
+  // If the witness method mutates self, we cannot replace self.
+  if (Apply.getOrigCalleeType()->getSelfParameter().isIndirectMutating())
+    return nullptr;
+
+  SILValue Self = Apply.getSelfArgument();
+  auto *PD = WMI->getLookupProtocol();
+  ConcreteExistentialInfo CEI;
+  /// Determine concrete type and it's conformance.
+  if (!prepareNCheckSoleConformingTypePropagation(Apply, PD, Self, CEI))
+    return nullptr;
+
+  auto &M = Builder.getModule();
+  auto ConformanceRef =
+      M.getSwiftModule()->lookupConformance(CEI.ConcreteType, PD);
+  auto SelfConformance =
+      ProtocolConformanceRef(ConformanceRef.getValue().getConcrete());
+  CEI.ExistentialConformances =
+      ArrayRef<ProtocolConformanceRef>(SelfConformance);
+  CEI.ExistentialType = PD->getDeclaredType()->getCanonicalType();
+  CanGenericSignature ExistentialSig =
+      M.getASTContext().getExistentialSignature(CEI.ExistentialType,
+                                                M.getSwiftModule());
+  CEI.ExistentialSubs = SubstitutionMap::get(ExistentialSig, {CEI.ConcreteType},
+                                             CEI.ExistentialConformances);
+  SILBuilderContext BuilderCtx(Builder.getModule(), Builder.getTrackingList());
+  if (CEI.ConcreteType != WMI->getLookupType() ||
+      SelfConformance != WMI->getConformance()) {
+    SILBuilderWithScope WMIBuilder(WMI, BuilderCtx);
+    auto *NewWMI = WMIBuilder.createWitnessMethod(
+        WMI->getLoc(), CEI.ConcreteType, SelfConformance, WMI->getMember(),
+        WMI->getType());
+    MutableArrayRef<Operand> Operands =
+        Apply.getInstruction()->getAllOperands();
+    for (auto &Op : Operands) {
+      if (Op.get() == WMI)
+        Op.set(NewWMI);
+    }
+    if (WMI->use_empty())
+      eraseInstFromFunction(*WMI);
+  }
+  auto *NewAI = createApplyWithConcreteType(Apply, CEI, BuilderCtx);
```

Comment:
Create a new overload of `createApplyWithConcreteType()` so you can share this block of code between here and `SILCombiner::propagateConcreteTypeOfInitExistential()`